document/JS.bs

```diff
@@ -0,0 +1,935 @@
+
+Issue: In future versions, WebAssembly
+be loaded and run directly from an HTML `<script type='module'>` tag—and
+any other Web API that loads ES6 modules via URL—as part of
+[ES6 Module integration](https://github.com/WebAssembly/design/issues/1087).)
+
+Note: WebAssembly JS API declaration file for TypeScript can be found [here](https://github.com/01alchemist/webassembly-types/blob/master/webassembly.d.ts) which enable autocompletion and make TypeScript compiler happy.
+
+<h2 id="sample">Sample API usage</h2>
+
+<p><em>This section is non-normative.</em></p>
+
+Given `demo.was` (encoded to `demo.wasm`):
+
+```lisp
+(module
+    (import "js" "import1" (func $i1))
+    (import "js" "import2" (func $i2))
+    (func $main (call $i1))
+    (start $main)
+    (func (export "f") (call $i2))
+)
+```
+
+and the following JavaScript, run in a browser:
+
+```javascript
+var importObj = {js: {
+    import1: () => console.log("hello,"),
+    import2: () => console.log("world!")
+}};
+fetch('demo.wasm').then(response =>
+    response.arrayBuffer()
+).then(buffer =>
+    WebAssembly.instantiate(buffer, importObj)
+).then(({module, instance}) =>
+    instance.exports.f()
+);
+```
+
+<h2 id="webassembly-namespace">The WebAssembly Namespace</h2>
+
+<pre class="idl">
+typedef (ExportedFunction or double or
+         Memory or Table)
+    WebAssemblyRuntimeValue;
+
+typedef (record&lt;DOMString, WebAssemblyRuntimeValue&gt;) InstanceExportsMap;
+
+typedef record&lt;DOMString, InstanceExportsMap&gt; InstanceImportsMap;
+
+dictionary WebAssemblyInstantiatedSource {
+   required Module module;
+   required Instance instance;
+};
+
+[Exposed=(Window,Worker)]
+namespace WebAssembly {
+  boolean validate(BufferSource bytes);
+  Promise&lt;Module> compile(BufferSource bytes);
+
+  Promise&lt;WebAssemblyInstantiatedSource> instantiate(
+      BufferSource bytes, optional InstanceImportsMap importObject);
+
+  Promise&lt;Instance> instantiate(
+      Module moduleObject, optional InstanceImportsMap importObject);
+};
+</pre>
+
+<!--
+Should we include notes describing what the functions do, as the HTML spec does? It could look like this:
+
+Note:
+  WebAssembly.validate(|bytes|) synchronously validates bytes of WebAssembly, returning true if the validation was successful.
+  WebAssembly.compile(|bytes|) asynchronously validates and complies bytes of WebAssembly into a Module.
+  WebAssembly.instantiate(|bytes|, |importObject|) asynchronously compiles and instantiates a WebAssembly module from bytes of source.
+  The WebAssembly.instantiate(|moduleObject|, |importObject|) asynchronously instantiates a compiled module.
+-->
+
+<div algorithm>
+  A WebAssembly module AST |module| is said to be <dfn>valid</dfn> if:
+    1. [=validate_module=](_module_) is not [=error=], and
+    1. The UTF-8 encoding of all <code>name</code> productions in the binary format, including in a custom <code>name</code> section if present, is valid.
```

Comment:
It's unclear what UTF-8 encoding being "invalid" would mean. In any case this should link to https://encoding.spec.whatwg.org/#utf-8-encode . Not that that algorithm never fails.

Reply:
The wording was a little unclear; I was getting at decoding, not encoding. Made a reference to your spec (since it seems like it doesn't paper over errors).